crates/circuit/src/operations.rs

```diff
@@ -167,6 +167,7 @@ pub trait Operation {
     fn definition(&self, params: &[Param]) -> Option<CircuitData>;
     fn standard_gate(&self) -> Option<StandardGate>;
     fn directive(&self) -> bool;
+    fn inverse(&self, params: &[Param]) -> Option<(StandardGate, Vec<Param>)>;
```

Comment:
Does this work in the case of `PyGate`? I don't think there is a guarantee that a Python defined gate's inverse will be a `StandardGate`. I actually think it is pretty unlikely in practice, since if a custom gate were expressible as the inverse of a `StandardGate` it'd probably just be a `StandardGate`. I think you can just do:

```suggestion
    fn inverse(&self, params: &[Param]) -> Option<(Self, Vec<Param>)>;
```
then for `impl Operation for StandardGate` `.inverse()` returns a `StandardGate` and for `impl Operator for PyGate` `.inverse()` returns a `PyGate`.

Reply:
In this case, it's unclear to me which type to put in line 279. Any type I try gives a mismatched type error

Reply:
We probably can't have `Operation::inverse` return anything referring to `Self`, because that's unimplementable by the implementors of `Operation` that don't have an owned type (e.g. `OperationRef`).  Even within that, it's maybe suboptimal because not all `StandardGate`s would be able to return an inverse, even though they can be inverted (e.g. `DCX`).

The most general would potentially be to have `Operation::inverse` return some iterator type over (potentially) `PackedInstruction` (like `CircuitData`), but in practice that's probably not super performant for many simple cases.  Maybe in the near term, it's enough just to have `StandardGate::try_inverse` on `StandardGate` only?

Reply:
There is a problem in implementing `inverse` in `impl<'a> Operation for OperationRef<'a>` (line 279) that cause a mismatched type error. After talking with @jakelishman it seems that the best solution would be to implement `inverse` only for `StandadGate` and not as part of the `Operation` trait.

Reply:
I'm a bit confused now. I can add `inverse` as a stand-alone function (only for `StandardGate`) - but then how can I test it from Python ?

```
pub fn inverse(gate:StandardGate, params: &[Param]) -> Option<(StandardGate, SmallVec<[Param; 3]>)> {
    match gate {
        StandardGate::HGate => Some((StandardGate::HGate, smallvec![])),
        StandardGate::IGate => Some((StandardGate::IGate, smallvec![])),
        ...
```

Reply:
So I replaced `inverse` by `try_inverse` and the function for `PyGate` will return `None`.